File: web/src/sim/compile.ts
/**
 * Rule-based compiler from a parsed note to an InstructionSet document
 * (port of expertloop/compiler/compile.py). Every step carries at least one citation:
 * the line range of the note item it came from plus any source references found on it.
 */

import { type NoteItem, type ParsedNote, type SourceRef, parseNote, sectionItems } from "./parser";

export interface Citation {
  note_id: number | null;
  line_start: number;
  line_end: number;
  source_kind?: string;
  source_ref?: string;
  source_id?: number;
  source_hash?: string;
}

export interface DecisionRule {
  condition: string;
  then: string;
  halts: boolean;
  citations?: Citation[];
}

export interface Step {
  id: string;
  order: number;
  action: string;
  condition: string | null;
  halts: boolean;
  tool: string | null;
  decision_rules: DecisionRule[];
  forbidden: string[];
  expected_outcome: string | null;
  citations: Citation[];
}

export interface CitedText {
  text: string;
  citations: Citation[];
}

export interface InstructionDocument {
  name: string;
  title: string;
  preconditions: CitedText[];
  steps: Step[];
  decision_rules: DecisionRule[];
  tools: string[];
  outcomes: CitedText[];
  forbidden_actions: CitedText[];
  sources: SourceRef[];
  agent_prompt: string;
}

const TOOL_RE = /\b(?:in|via|using|through|open|call|from)\s+([A-Z][A-Za-z0-9]+(?:\s[A-Z][A-Za-z0-9]+)?)/;
const RULE_RE = /^\s*(?:if|when)\s+(.+?)\s*(?:,|then)\s*(.+)$/i;
const FORBIDDEN_RE = /\b(?:never|do not|don't|must not)\s+(.+)/i;
const OUTCOME_RE = /\b(?:so that|until|expected:|result:)\s*(.+)$/i;
export const STOP_WORDS = ["stop", "halt", "escalate", "do not proceed", "hand off", "hand it off", "pause"];

function rstripDot(value: string): string {
  return value.replace(/\.+$/, "");
}

function hasStopWord(text: string): boolean {
  const lowered = text.toLowerCase();
  return STOP_WORDS.some((word) => lowered.includes(word));
}

function citation(item: NoteItem, noteId: number | null, ref?: SourceRef): Citation {
  const cite: Citation = { note_id: noteId, line_start: item.line_start, line_end: item.line_end };
  if (ref) {
    cite.source_kind = ref.kind;
    cite.source_ref = ref.ref;
  }
  return cite;
}

function citations(item: NoteItem, noteId: number | null): Citation[] {
  return [citation(item, noteId), ...item.sources.map((ref) => citation(item, noteId, ref))];
}

function firstLine(text: string): string {
  return text.split("\n", 1)[0].trim();
}

function detectTool(text: string, knownTools: string[]): string | null {
  const lowered = text.toLowerCase();
  for (const tool of knownTools) {
    if (lowered.includes(tool.toLowerCase())) return tool;
  }
  const match = TOOL_RE.exec(text);
  return match ? match[1] : null;
}

interface SplitResult {
  action: string;
  condition: string | null;
  rules: DecisionRule[];
  forbidden: string[];
  outcome: string | null;
}

/**
 * Separate a step into action, gating condition, decision rules, forbidden clauses and outcome.
 * A step whose first line is itself "if X, Y" is conditional: it runs only when X holds and its
 * action is Y. Rules on later lines are decision rules evaluated before the action.
 */
function splitRules(text: string): SplitResult {
  const actionLines: string[] = [];
  let condition: string | null = null;
  const rules: DecisionRule[] = [];
  const forbidden: string[] = [];
  let outcome: string | null = null;
  text.split("\n").forEach((rawLine, index) => {
    let line = rawLine;
    const rule = RULE_RE.exec(line);
    if (rule && index === 0) {
      condition = rule[1].trim();
      line = rule[2].trim();
    } else if (rule) {
      const then = rstripDot(rule[2].trim());
      rules.push({ condition: rule[1].trim(), then, halts: hasStopWord(then) });
      return;
    }
    const banned = FORBIDDEN_RE.exec(line);
    if (banned) {
      forbidden.push(rstripDot(banned[1].trim()));
      return;
    }
    const result = OUTCOME_RE.exec(line);
    if (result && outcome === null) {
      const captured = result[1].trim();
      const split = captured.indexOf(". ");
      let rest = "";
      if (split >= 0) {
        outcome = captured.slice(0, split);
        rest = captured.slice(split + 2);
      } else {
        outcome = captured;
      }
      outcome = rstripDot(outcome);
      line = line.slice(0, result.index).trim().replace(/,+$/, "");
      if (line) actionLines.push(line);
      if (rest) actionLines.push(rest.trim());
      return;
    }
    actionLines.push(line.trim());
  });
  const action = actionLines.filter(Boolean).join(" ").trim();
  return { action, condition, rules, forbidden, outcome };
}

export function compileParsed(parsed: ParsedNote, noteId: number | null, name: string): InstructionDocument {
  const tools = sectionItems(parsed, "tools").map((item) => rstripDot(firstLine(item.text)));
  const steps: Step[] = [];
  const globalRules: DecisionRule[] = [];
  const forbiddenActions: CitedText[] = [];
  const preconditions = sectionItems(parsed, "preconditions").map((item) => ({
    text: rstripDot(firstLine(item.text)),
    citations: citations(item, noteId),
  }));
  const outcomes = sectionItems(parsed, "outcomes").map((item) => ({
    text: rstripDot(firstLine(item.text)),
    citations: citations(item, noteId),
  }));

  for (const item of sectionItems(parsed, "decision_rules")) {
    const rule = RULE_RE.exec(firstLine(item.text));
    if (rule) {
      const then = rstripDot(rule[2].trim());
      globalRules.push({
        condition: rule[1].trim(),
        then,
        halts: hasStopWord(then),
        citations: citations(item, noteId),
      });
    } else {
      globalRules.push({
        condition: rstripDot(firstLine(item.text)),
        then: "apply rule",
        halts: false,
        citations: citations(item, noteId),
      });
    }
  }

  for (const item of sectionItems(parsed, "forbidden")) {
    const banned = FORBIDDEN_RE.exec(item.text);
    const text = banned ? banned[1] : firstLine(item.text);
    forbiddenActions.push({ text: rstripDot(text.trim()), citations: citations(item, noteId) });
  }

  sectionItems(parsed, "steps").forEach((item, i) => {
    const index = i + 1;
    const { condition, rules, forbidden, outcome, action: split } = splitRules(item.text);
    const action = split || firstLine(item.text);
    steps.push({
      id: `s${index}`,
      order: index,
      action: rstripDot(action),
      condition,
      halts: hasStopWord(action),
      tool: detectTool(item.text, tools),
      decision_rules: rules,
      forbidden,
      expected_outcome: outcome,
      citations: citations(item, noteId),
    });
    for (const text of forbidden) {
      forbiddenActions.push({ text, citations: citations(item, noteId) });
    }
  });

  const document: InstructionDocument = {
    name,
    title: parsed.title ?? name,
    preconditions,
    steps,
    decision_rules: globalRules,
    tools,
    outcomes,
    forbidden_actions: forbiddenActions,
    sources: parsed.sources.map((ref) => ({ kind: ref.kind, ref: ref.ref })),
    agent_prompt: "",
  };
  document.agent_prompt = renderPrompt(document);
  return document;
}

export function compileNote(body: string, noteId: number | null = null, name?: string): InstructionDocument {
  const parsed = parseNote(body);
  return compileParsed(parsed, noteId, name ?? parsed.title ?? "untitled");
}

/** Render the instruction set as the text an agent will receive. */
export function renderPrompt(document: InstructionDocument): string {
  const lines: string[] = [`# ${document.title}`, ""];
  if (document.preconditions.length) {
    lines.push("Before starting, confirm:");
    for (const p of document.preconditions) lines.push(`- ${p.text}`);
    lines.push("");
  }
  lines.push("Follow these steps in order:");
  for (const step of document.steps) {
    const tool = step.tool ? ` (tool: ${step.tool})` : "";
    const gate = step.condition ? `Only if ${step.condition}: ` : "";
    lines.push(`${step.order}. ${gate}${step.action}${tool}`);
    for (const rule of step.decision_rules) lines.push(`   - if ${rule.condition}: ${rule.then}`);
    if (step.expected_outcome) lines.push(`   - expected: ${step.expected_outcome}`);
  }
  if (document.decision_rules.length) {
    lines.push("", "Decision rules:");
    for (const r of document.decision_rules) lines.push(`- if ${r.condition}: ${r.then}`);
  }
  if (document.forbidden_actions.length) {
    lines.push("", "Never:");
    for (const f of document.forbidden_actions) lines.push(`- ${f.text}`);
  }
  if (document.outcomes.length) {
    lines.push("", "Done when:");
    for (const o of document.outcomes) lines.push(`- ${o.text}`);
  }
  return lines.join("\n");
}

export interface Coverage {
  steps: number;
  cited_steps: number;
  citations: number;
  coverage: number;
}

export function citationCoverage(document: InstructionDocument): Coverage {
  const steps = document.steps;
  const cited = steps.filter((s) => s.citations.length > 0);
  const total = steps.length;
  return {
    steps: total,
    cited_steps: cited.length,
    citations: steps.reduce((sum, s) => sum + s.citations.length, 0),
    coverage: total ? cited.length / total : 1,
  };
}

/** Return a list of problems; an empty list means the document is acceptable. */
export function validateDocument(document: Partial<InstructionDocument>): string[] {
  const problems: string[] = [];
  for (const key of ["title", "steps", "preconditions", "decision_rules", "forbidden_actions"] as const) {
    if (!(key in document)) problems.push(`missing field: ${key}`);
  }
  const steps = document.steps ?? [];
  for (const step of steps) {
    if (!step.id) problems.push("step without id");
    if (!step.action) problems.push(`step ${step.id} has no action`);
    if (!step.citations || step.citations.length === 0) problems.push(`step ${step.id} has no citations`);
  }
  const ids = steps.map((s) => s.id);
  if (ids.length !== new Set(ids).size) problems.push("duplicate step ids");
  return problems;
}

File: expertloop/compiler/compile.py
"""Rule-based compiler from a parsed note to an ``InstructionSet`` document.

The output is a plain dictionary (stored as JSONB) so that edits can be diffed and
versioned without an ORM round trip. Every step carries at least one citation: the
line range of the note item it came from plus any source references found on it.
"""

from __future__ import annotations

import re
from typing import Any

from expertloop.compiler.parser import NoteItem, ParsedNote, SourceRef, parse_note

TOOL_RE = re.compile(r"\b(?:in|via|using|through|open|call|from)\s+([A-Z][A-Za-z0-9]+(?:\s[A-Z][A-Za-z0-9]+)?)")
RULE_RE = re.compile(r"^\s*(?:if|when)\s+(.+?)\s*(?:,|then)\s*(.+)$", re.IGNORECASE)
FORBIDDEN_RE = re.compile(r"\b(?:never|do not|don't|must not)\s+(.+)", re.IGNORECASE)
OUTCOME_RE = re.compile(r"\b(?:so that|until|expected:|result:)\s*(.+)$", re.IGNORECASE)
STOP_WORDS = ("stop", "halt", "escalate", "do not proceed", "hand off", "hand it off", "pause")


def _citation(item: NoteItem, note_id: int | None, ref: SourceRef | None = None) -> dict[str, Any]:
    cite: dict[str, Any] = {
        "note_id": note_id,
        "line_start": item.line_start,
        "line_end": item.line_end,
    }
    if ref is not None:
        cite["source_kind"] = ref.kind
        cite["source_ref"] = ref.ref
    return cite


def _citations(item: NoteItem, note_id: int | None) -> list[dict[str, Any]]:
    cites = [_citation(item, note_id)]
    cites.extend(_citation(item, note_id, ref) for ref in item.sources)
    return cites


def _first_line(text: str) -> str:
    return text.split("\n", 1)[0].strip()


def _detect_tool(text: str, known_tools: list[str]) -> str | None:
    lowered = text.lower()
    for tool in known_tools:
        if tool.lower() in lowered:
            return tool
    match = TOOL_RE.search(text)
    return match.group(1) if match else None


def _split_rules(text: str) -> tuple[str, list[dict[str, str]], list[str], str | None]:
    """Separate a step into its action, embedded decision rules, forbidden clauses, outcome."""
    action_lines: list[str] = []
    rules: list[dict[str, str]] = []
    forbidden: list[str] = []
    outcome: str | None = None
    for line in text.split("\n"):
        rule = RULE_RE.match(line)
        if rule:
            then = rule.group(2).strip().rstrip(".")
            rules.append(
                {
                    "condition": rule.group(1).strip(),
                    "then": then,
                    "halts": any(word in then.lower() for word in STOP_WORDS),
                }
            )
            continue
        banned = FORBIDDEN_RE.search(line)
        if banned:
            forbidden.append(banned.group(1).strip().rstrip("."))
            continue
        result = OUTCOME_RE.search(line)
        if result and outcome is None:
            outcome = result.group(1).strip().rstrip(".")
            line = line[: result.start()].strip().rstrip(",")
            if line:
                action_lines.append(line)
            continue
        action_lines.append(line.strip())
    action = " ".join(part for part in action_lines if part).strip()
    return action, rules, forbidden, outcome


def compile_parsed(parsed: ParsedNote, note_id: int | None, name: str) -> dict[str, Any]:
    tools = [_first_line(item.text).rstrip(".") for item in parsed.section("tools")]
    steps: list[dict[str, Any]] = []
    global_rules: list[dict[str, Any]] = []
    forbidden_actions: list[dict[str, Any]] = []
    preconditions = [
        {"text": _first_line(item.text).rstrip("."), "citations": _citations(item, note_id)}
        for item in parsed.section("preconditions")
    ]
    outcomes = [
        {"text": _first_line(item.text).rstrip("."), "citations": _citations(item, note_id)}
        for item in parsed.section("outcomes")
    ]

    for item in parsed.section("decision_rules"):
        rule = RULE_RE.match(_first_line(item.text))
        if rule:
            then = rule.group(2).strip().rstrip(".")
            global_rules.append(
                {
                    "condition": rule.group(1).strip(),
                    "then": then,
                    "halts": any(word in then.lower() for word in STOP_WORDS),
                    "citations": _citations(item, note_id),
                }
            )
        else:
            global_rules.append(
                {
                    "condition": _first_line(item.text).rstrip("."),
                    "then": "apply rule",
                    "halts": False,
                    "citations": _citations(item, note_id),
                }
            )

    for item in parsed.section("forbidden"):
        banned = FORBIDDEN_RE.search(item.text)
        text = banned.group(1) if banned else _first_line(item.text)
        forbidden_actions.append(
            {"text": text.strip().rstrip("."), "citations": _citations(item, note_id)}
        )

    for index, item in enumerate(parsed.section("steps"), start=1):
        action, rules, banned, outcome = _split_rules(item.text)
        if not action:
            action = _first_line(item.text)
        step = {
            "id": f"s{index}",
            "order": index,
            "action": action.rstrip("."),
            "tool": _detect_tool(item.text, tools),
            "decision_rules": rules,
            "forbidden": banned,
            "expected_outcome": outcome,
            "citations": _citations(item, note_id),
        }
        steps.append(step)
        for text in banned:
            forbidden_actions.append({"text": text, "citations": _citations(item, note_id)})

    document = {
        "name": name,
        "title": parsed.title or name,
        "preconditions": preconditions,
        "steps": steps,
        "decision_rules": global_rules,
        "tools": tools,
        "outcomes": outcomes,
        "forbidden_actions": forbidden_actions,
        "sources": [{"kind": ref.kind, "ref": ref.ref} for ref in parsed.sources],
    }
    document["agent_prompt"] = render_prompt(document)
    return document


def compile_note(body: str, note_id: int | None = None, name: str | None = None) -> dict[str, Any]:
    parsed = parse_note(body)
    return compile_parsed(parsed, note_id, name or parsed.title or "untitled")


def render_prompt(document: dict[str, Any]) -> str:
    """Render the instruction set as the text an agent will receive."""
    lines = [f"# {document['title']}", ""]
    if document["preconditions"]:
        lines.append("Before starting, confirm:")
        lines.extend(f"- {p['text']}" for p in document["preconditions"])
        lines.append("")
    lines.append("Follow these steps in order:")
    for step in document["steps"]:
        tool = f" (tool: {step['tool']})" if step.get("tool") else ""
        lines.append(f"{step['order']}. {step['action']}{tool}")
        for rule in step["decision_rules"]:
            lines.append(f"   - if {rule['condition']}: {rule['then']}")
        if step.get("expected_outcome"):
            lines.append(f"   - expected: {step['expected_outcome']}")
    if document["decision_rules"]:
        lines.append("")
        lines.append("Decision rules:")
        lines.extend(f"- if {r['condition']}: {r['then']}" for r in document["decision_rules"])
    if document["forbidden_actions"]:
        lines.append("")
        lines.append("Never:")
        lines.extend(f"- {f['text']}" for f in document["forbidden_actions"])
    if document["outcomes"]:
        lines.append("")
        lines.append("Done when:")
        lines.extend(f"- {o['text']}" for o in document["outcomes"])
    return "\n".join(lines)


def citation_coverage(document: dict[str, Any]) -> dict[str, Any]:
    steps = document.get("steps", [])
    cited = [s for s in steps if s.get("citations")]
    total = len(steps)
    return {
        "steps": total,
        "cited_steps": len(cited),
        "citations": sum(len(s.get("citations", [])) for s in steps),
        "coverage": (len(cited) / total) if total else 1.0,
    }


def validate_document(document: dict[str, Any]) -> list[str]:
    """Return a list of problems; an empty list means the document is acceptable."""
    problems: list[str] = []
    for key in ("title", "steps", "preconditions", "decision_rules", "forbidden_actions"):
        if key not in document:
            problems.append(f"missing field: {key}")
    for step in document.get("steps", []):
        if not step.get("id"):
            problems.append("step without id")
        if not step.get("action"):
            problems.append(f"step {step.get('id')} has no action")
        if not step.get("citations"):
            problems.append(f"step {step.get('id')} has no citations")
    ids = [s.get("id") for s in document.get("steps", [])]
    if len(ids) != len(set(ids)):
        problems.append("duplicate step ids")
    return problems
